### PR TITLE
Created ReadOnlySequence tests and fixed issues

### PR DESCRIPTION
**Pull request by AlexRadch**

### CODE REVIEW

File: src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs [hunk 1]
```diff
@@ -142,66 +142,66 @@ internal ReadOnlyMemory<T> GetFirstBuffer(in SequencePosition start, in Sequence
...
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal SequencePosition Seek(in SequencePosition start, in SequencePosition end, long count)
+        internal SequencePosition Seek(in SequencePosition start, in SequencePosition end, long offset)
         {
             GetTypeAndIndices(start.GetInteger(), end.GetInteger(), out SequenceType type, out int startIndex, out int endIndex);
 
             object startObject = start.GetObject();
-            object endObject = end.GetObject();
+            object endObject;
```
Comment (ahsonkhan) [inline src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs]: Why was this change necessary?

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs [hunk 2]
```diff
@@ -13,100 +13,115 @@ namespace System.Memory.Tests
...
-            SequencePosition c1 = buffer.GetPosition(25); // segment 1 index 75
-            SequencePosition c2 = buffer.GetPosition(55); // segment 2 index 5
+            var c1 = buffer.GetPosition(25); // segment 1 index 75
+            var c2 = buffer.GetPosition(55); // segment 2 index 5
+            Assert.Equal(5, c2.GetInteger());
+            Assert.Equal(bufferSegment2, c2.GetObject());
+            Assert.Equal(c2, buffer.GetPosition(55, buffer.Start));
 
-            ReadOnlySequence<byte> sliced = buffer.Slice(c1, c2);
+            var sliced = buffer.Slice(c1, c2);
```
Comment (ahsonkhan) [inline src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: nit: don't use var here, state the type explicitly

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs [hunk 3]
```diff
@@ -303,73 +412,205 @@ public void EnumerableStopsAtEndWhenEndIsFirstByteOfEmpty()
...
+            Assert.True(memory.IsEmpty);
+
+            Assert.False(buffer.TryGet(ref c1, out memory, false));
+            Assert.Equal(null, c1.GetObject());
+            Assert.True(memory.IsEmpty);
+
+            Assert.False(buffer.TryGet(ref c1, out memory, true));
+            Assert.Equal(null, c1.GetObject());
+            Assert.True(memory.IsEmpty);
+            //Assert.Throws<ArgumentOutOfRangeException>(() => buffer.TryGet(ref c1, out memory, false));
```
Comment (ahsonkhan) [inline src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: Why is this commented out?

**Comment by ahsonkhan** [comment]
@AlexRadch, how does code coverage look now? Can you measure please and share?

File: src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs (hunk 1 above)
Comment (AlexRadch) [reply src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs]: It will generate more optimized code.

Comment (AlexRadch) [reply src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs]: Also I renamed `count` to offset because it produce ArgumentException with wrong name for GetPosition methods.

File: src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs [hunk 4]
```diff
@@ -16,25 +16,14 @@ namespace System.Buffers
...
+                data = default;
+                return false;
             }
-
-            int length = endIndex - startIndex;
-            object endObject = end.GetObject();
-
-            if (type != SequenceType.MultiSegment && type != SequenceType.Empty && startObject != endObject)
-                ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
+            GetTypeAndIndices(start.GetInteger(), end.GetInteger(), out var type, out var startIndex, out var endIndex);
```
Comment (ahsonkhan) [inline src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs]: nit: don't use var here. explicitly state the type. only use var where you can statically infer what the type is (for instance if the right hand side is `new ...`.

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs [hunk 5]
```diff
@@ -203,62 +292,81 @@ public void TryGetStopsAtEnd()
...
-        [Fact]
-        public void TryGetStopsAtEndWhenEndIsFirstByteOfEmpty()
+        [Theory]
+        [InlineData(0)]
+        [InlineData(long.MaxValue - 50)]
+        [InlineData(-50)]
+        public void TryGetStopsAtEndWhenEndIsFirstByteOfEmpty(long startIndex)
         {
-            var bufferSegment1 = new BufferSegment<byte>(new byte[100]);
+            new BufferSegment<byte> bufferSegment1 = new BufferSegment<byte>(new byte[100], startIndex);
```
Comment (benaadams) [inline src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: Doesn't compile, line can't start with `new `

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs [hunk 6]
```diff
@@ -303,150 +417,292 @@ public void EnumerableStopsAtEndWhenEndIsFirstByteOfEmpty()
...
+            Assert.Throws<ArgumentNullException>("array", () => new ReadOnlySequence<byte>((byte[])null, 4, 2));
         }
 
         [Fact]
         public void Ctor_OwnedMemory_Offset()
         {
-            var ownedMemory = new CustomMemoryForTest<byte>(new byte[] { 1, 2, 3, 4, 5 }, 0, 5);
-            var buffer = new ReadOnlySequence<byte>(ownedMemory, 2, 3);
+            CustomMemoryForTest<byte> ownedMemory = new CustomMemoryForTest<byte>(new byte[] { 1, 2, 3, 4, 5 }, 0, 5);
+            new ReadOnlySequence<byte> buffer = new ReadOnlySequence<byte>(ownedMemory, 2, 3);
```
Comment (benaadams) [inline src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: Doesn't compile, line can't start with `new`

File: src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs [hunk 7]
```diff
@@ -43,165 +32,111 @@ internal bool TryGetBuffer(in SequencePosition start, in SequencePosition end, o
...
+                    Debug.Assert(startObject is OwnedMemory<T>);
 
-                memory = (Unsafe.As<OwnedMemory<T>>(startObject)).Memory;
-            }
-            else if (typeof(T) == typeof(char) && type == SequenceType.String)
-            {
-                Debug.Assert(startObject is string);
+                    data = (Unsafe.As<OwnedMemory<T>>(startObject)).Memory;
+                }
+                else
```
Comment (benaadams) [inline src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs]: Still want 
```csharp
else if (typeof(T) == typeof(char)
```
As if `T` is not `char` e.g. `byte` the Jit will then remove the block entirely

**Comment by davidfowl** [comment]
@AlexRadch thanks for the PR, you made a bunch of changes without a comment in the PR or any performance numbers, are you working on those?

**Comment by AlexRadch** [comment]
@davidfowl Yes, I found many optimization issues in ReadOnlySequence so I now working with ReadOnlySequence optimization.

**Comment by AlexRadch** [comment]
I updated ReadOnlySequence tests, removed test fails and optimized some methods.
@davidfowl How to make performance tests for updated methods?

**Comment by pakrym** [comment]
Can you please split this PR into a separate ones targeting specific problems/improvements? It's very hard to review when both implementation and tests are changed in such large quantities.

**Comment by AlexRadch** [comment]
@pakrym I will try ;)

**Comment by ahsonkhan** [comment]
@AlexRadch, what's your plan for this PR? Are there any changes in here that aren't covered in the other PRs: https://github.com/dotnet/corefx/pull/28760 / https://github.com/dotnet/corefx/pull/28702?

**Comment by AlexRadch** [comment]
@ahsonkhan I think I close this PR because I am making similar changes by separated parts.
